storm-core/src/jvm/org/apache/storm/zookeeper/Zookeeper.java

```diff
@@ -0,0 +1,355 @@
+            @Override
+            public void eventReceived(CuratorFramework _fk, CuratorEvent e) throws Exception {
+                if (e.getType().equals(CuratorEventType.WATCHED)) {
+                    WatchedEvent event = e.getWatchedEvent();
+
+                    watcher.execute(event.getState(), event.getType(), event.getPath());
+                }
+
+            }
+        });
+        fk.start();
+        return fk;
+    }
+
+    /**
+     * connect ZK, register Watch/unhandle Watch
+     *
+     * @return
+     */
+    public static CuratorFramework mkClient(Map conf, List<String> servers, Object port, String root, final WatcherCallBack watcher) {
+
+        return mkClient(conf, servers, port, root, watcher, null);
+    }
+
+    public static String createNode(CuratorFramework zk, String path, byte[] data, org.apache.zookeeper.CreateMode mode, List<ACL> acls)
+            throws RuntimeException {
+
+        String ret = null;
+        try {
+            String npath = Utils.normalizePath(path);
+            ret = zk.create().withMode(mode).withACL(acls).forPath(npath, data);
+        } catch (Exception e) {
+            throw Utils.wrapInRuntime(e);
+        }
+        return ret;
+    }
+
+    public static String createNode(CuratorFramework zk, String path, byte[] data, List<ACL> acls) throws RuntimeException {
+        return createNode(zk, path, data, org.apache.zookeeper.CreateMode.PERSISTENT, acls);
+    }
+
+    public static boolean existsNode(CuratorFramework zk, String path, boolean watch) throws RuntimeException {
+        Stat stat = null;
+        try {
+            if (watch) {
+                stat = zk.checkExists().watched().forPath(Utils.normalizePath(path));
+            } else {
+                stat = zk.checkExists().forPath(Utils.normalizePath(path));
+            }
+        } catch (Exception e) {
+            throw Utils.wrapInRuntime(e);
+        }
+
+        return stat != null;
+    }
+
+    public static void deleteNode(CuratorFramework zk, String path) throws RuntimeException {
+        try {
+            String npath = Utils.normalizePath(path);
+            if (existsNode(zk, npath, false)) {
+                zk.delete().deletingChildrenIfNeeded().forPath(Utils.normalizePath(path));
+            }
+
+        } catch (KeeperException.NoNodeException e) {
+            LOG.info("exception", e);
+        } catch (Exception e) {
+            throw Utils.wrapInRuntime(e);
+        }
+    }
+
+    public static void mkdirs(CuratorFramework zk, String path, List<ACL> acls) throws RuntimeException {
+
+        String npath = Utils.normalizePath(path);
+        if (npath.equals("/")) {
+            return;
+        }
+        if (existsNode(zk, npath, false)) {
+            return;
+        }
+        byte[] byteArray = new byte[1];
+        byteArray[0] = (byte) 7;
+        createNode(zk, npath, byteArray, org.apache.zookeeper.CreateMode.PERSISTENT, acls);
```

Comment:
In this place we do need to do something like the try-cause that was in the original code.  There is a race condition where multiple clients can be doing a mkdirs.  If that happens we need to ignore the NodeExistsException from createNode, but createNode will have wrapped it in a RuntimeException.  If you want to avoid the wrapping/unwrapping and just catch the exception yourself without unwrapping it you could do the following instead.

```
try {
    zk.create().withMode(CreateMode.PERSISTENT).withACL(acls).forPath(npath, byteArray);
} catch (KeeperException.NodeExistsException ignored) {
    // this can happen when multiple clients doing mkdir at same time
}
```